ddtrace/internal/_encoding.pyx

```diff
@@ -118,14 +122,142 @@ cdef inline int pack_text(msgpack_packer *pk, object text):
+
+
+cdef class ListStringTable(StringTable):
+    cdef list _list
+
+    def __init__(self):
+        self._list = []
+        super(ListStringTable, self).__init__()
+
+    cdef insert(self, object string):
+        PyList_Append(self._list, string)
+
+    def __iter__(self):
+        return iter(self._list)
+
+
+cdef class MsgpackStringTable(StringTable):
+    cdef msgpack_packer pk
+    cdef int max_size
+    cdef int _sp
+    cdef object _lock
+
+    def __init__(self, max_size):
+        self.pk.buf = <char*> PyMem_Malloc(max_size)
+        if self.pk.buf == NULL:
+            raise MemoryError("Unable to allocate internal buffer.")
+        self.max_size = self.pk.buf_size = max_size
+        self.pk.length = 6
+        self._sp = 0
+        self._lock = threading.Lock()
+        super(MsgpackStringTable, self).__init__()
+
+    def __dealloc__(self):
+        PyMem_Free(self.pk.buf)
+        self.pk.buf = NULL
+
+    cdef insert(self, object string):
+        cdef int ret
+
+        ret = pack_text(&self.pk, string)
+        if ret != 0:
+            raise RuntimeError("Failed to add string to msgpack string table")
+
+    cdef savepoint(self):
+        self._sp = self.pk.length
+
+    cdef rollback(self):
+        if self._sp > 0:
+            self.pk.length = self._sp
+
+    cdef get_bytes(self):
+        cdef stdint.uint32_t l = self._next_id
+        cdef int offset = 6 - array_prefix_size(l)
+        cdef int old_pos = self.pk.length
+        
+        with self._lock:
+            # Update table size prefix
+            self.pk.length = offset
+            msgpack_pack_array(&self.pk, l)
+            # Add root array size prefix
+            self.pk.length = offset = offset - 1
+            msgpack_pack_array(&self.pk, 2)
+            self.pk.length = old_pos
+        
+        return PyBytes_FromStringAndSize(self.pk.buf + offset, self.pk.length - offset)
+
+    @property
+    def size(self):
+        return self.pk.length - 5 + array_prefix_size(self._next_id)
+
+    cdef append_raw(self, object src, Py_ssize_t size):
+        cdef int res
+        with self._lock:
+            assert self.size + size <= self.max_size
+            res = msgpack_pack_raw_body(&self.pk, <char *>PyLong_AsLong(src), size)
+            if res != 0:
+                raise RuntimeError("Failed to append raw bytes to msgpack string table")
+
+
+    cpdef flush(self):
+        try:
+            return self.get_bytes()
+        finally:
+            self.reset()
+            self.pk.length = 6
+            self._sp = 0
+
+
 cdef class BufferedEncoder(object):
     content_type: str = None
 
-    cdef public int max_size
-    cdef public int max_item_size
+    cdef public Py_ssize_t max_size
+    cdef public Py_ssize_t max_item_size
     cdef object _lock
 
-    def __cinit__(self, max_size, max_item_size):
+    def __cinit__(self, size_t max_size, size_t max_item_size):
         self.max_size = max_size
```

Comment:
Isn't the conversion from `size_t` to `Py_ssize_t` undefined if `max_size` doesn't fit in signed `Py_ssize_t`?

Reply:
Ah well 👀 . I thought I had changed all these types to `size_t`!